tests/src/JIT/HardwareIntrinsics/X86/Shared/Program.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+
+using System;
+
+namespace HardwareIntrinsicTest
+{
+    public static class Program
+    {
+        private const int Pass = 100;
+        private const int Fail = 0;
+
+        public static unsafe int Main(string[] args)
+        {
+            try
+            {
+                var test = new Test();
+
+                if (test.IsSupported)
```

Comment:
Split of tests into different calling patterns is very good, however, it could be more convenient to use Func<T> or similar to pass tests into main. This would allow for greater flexibility, since individual test would not have to implement all functions but only pass those which are relevant for them.

Reply:
The current functions represent the functionality that all tests should probably implement, in the most generic sense.

They ensure that the tests are functionally correct (for the various modes required) and that any new codegen works as intended.

There may be some tests which differ slightly, which is the purpose of the `ValidateOther`. For tests that differ more significantly, a more custom Main method is fine and would be expected.

Reply:
> The current functions represent the functionality that all tests should probably implement

Yes it is most probably correct but my intention was to suggest using more flexible pattern in discovering and running tests what would lift limitations on test implementation and decouple `Program` from `Test`. For instance every test method should implement `Func<bool>[] GetTests()` method which will return set of tests which will be run by `Main` function and some of the tests could be mandatory.

Reply:
And only `IsSupported` should be implemented directly by every test.

Reply:
> For instance every test method should implement Func<bool>[] GetTests() method

Fair enough. We should probably wait and see if moving to use a proper unit testing is viable, however (there may be some actual reason why not for CoreCLR)